### ISSUE TITLE
Update core to Backdrop CSS Coding standards

### ISSUE THREAD

**Issue by opi** [backdrop/backdrop-issues#2785]
BackdropCMS provides some guidelines about [CSS Coding Standards](https://api.backdropcms.org/css-standards) but a lot of core files does not respect them.

## Tools

- [StyleLint](https://stylelint.io/) with custom [configuration file](https://github.com/backdrop/backdrop-issues/files/1207934/stylelintrc.json.txt) and [ignore list](https://github.com/backdrop/backdrop-issues/files/1207933/stylelint-ignore.txt) (I don't want to patch third party stylesheets). Linting command is: `stylelint "core/**/*.css" --ignore-path stylelint-ignore.txt --config stylelintrc.json`.

- `find core/ -type f -not -path core/misc/ckeditor -not -path core/misc/ui -not -path core/misc/smartmenus -not -path core/module/simpletest  -name '*.css' -exec sed -i -E 's/:(\S+);/: \1;/g' {} \;` to add a whitespace between `property` and `value`.

- `find core/ -type f -not -path core/misc/ckeditor -not -path core/misc/ui -not -path core/misc/smartmenus -not -path core/module/simpletest  -name '*.css' -exec sed -i -E 's@#([A-F]{3,6});@#\L\1;@g' {} \;` to lowercase hex values.

- `find core/ -type f -not -path core/misc/ckeditor -not -path core/misc/ui -not -path core/misc/smartmenus -not -path core/module/simpletest -name '*.css' -exec sed -i -E 's@#([0-9a-fA-F])\1([0-9a-fA-F])\2([0-9a-fA-F])\3;@#\1\2\3;@g' {} \;` to use shorthand hex values.

---

Related:
* https://github.com/backdrop/backdrop-issues/issues/5245
* https://github.com/backdrop/backdrop-issues/issues/5568

**Comment by jenlampton** [comment]
Thanks @opi!  Feel free to submit PRs for any/all of these changes :)

**Comment by opi** [comment]
@jenlampton there is a PR already ;)

**Comment by jenlampton** [comment]
Yes, I see it there now. Not sure how I missed that last time!  Thank you @opi! (I will now add the relevant tags...)

**Comment by quicksketch** [comment]
Amazing @opi! Love it! We might consider adding `StyleLint` to our build process to validate code style.

One issue I saw repeated in the CSS (not a very big deal) is that it adds empty lines before comments, even if that is the first rule in the selector. e.g.

```diff
 fieldset {
+
   /* Webkit/Blink fix for mobile, see stackoverflow.com/questions/17408815 */
   min-width: 0;
   margin-bottom: 1em;
   padding: 0.5em;
 }
```

That new line is unnecessary. It makes sense in some other cases (but not all). Could we remove that rule and reroll the PR?

However this PR does SO MUCH more good than harm, if it's tedious to reroll I'd like to merge it anyway.

**Comment by quicksketch** [comment]
Actually the new-lines added throughout Basis are quite excessive. We need these to be fixed before merging.

**Comment by opi** [comment]
> Actually the new-lines added throughout Basis are quite excessive. We need these to be fixed before merging.

Do you mean that you agree with https://github.com/backdrop/backdrop/pull/1946/files#diff-ac5dbf064787a2cf66646b05c3f17c19R9, but not https://github.com/backdrop/backdrop/pull/1946/files#diff-ac5dbf064787a2cf66646b05c3f17c19R181 ?

**Comment by BWPanda** [comment]
I'm planning to make a few PRs related to this issue:

- Adding Stylelint rules to core
- Fixing existing CSS standards issues in core
- Setting up a GitHub action to check any new CSS added/changed in PRs

**Comment by BWPanda** [comment]
Here's a PR that adds the Stylelint rules: https://github.com/backdrop/backdrop/pull/3516

**Comment by BWPanda** [comment]
Here's the PR to fix existing issues in core CSS: https://github.com/backdrop/backdrop/pull/3517

Based on the stylelint rules in the above PR.

**Comment by BWPanda** [comment]
Once we can lock in the set of rules we check against, I'll *then* work on implementing a GitHub action to automate this in future.

**Comment by jenlampton** [comment]
Nice! :)

How would the action work? Would it mark issues as NW for all PRs that
don't pass?

(I was unable to make the issue / PR relationship work last te I tried
this. Curious to see how you get that going)

**Comment by BWPanda** [comment]
@jenlampton I haven't tried the GHA part yet, but I imagine it'll work similar to how the current tests work - just displays an error for the CSS standards test in the PR, then it's up to the author/reviewer to mark the related issue as 'needs work' manually.

But I can certainly see if I can get all that to happen automatically instead, if that's preferred...

**Comment by jenlampton** [comment]
Failing a test would be enough :) 
The problem I ran into was that the PRs and Issues were in different projects and none of the APIs seemed to support that.

**Comment by klonos** [comment]
Thanks so much for working on automating this as much as possible @BWPanda 🙏 ...and wow! ...I hadn't realized that our code had so many (obvious) standard  violations/inconsistencies.

**Comment by alanmels** [comment]
Applauds and thank you all!

**Comment by indigoxela** [comment]
A word of warning: changing color definitions from long to short, for instance ` background: #dddddd;` to `background: #ddd;` will have negative side effects. The color module will not be able to replace these values, as it relies on the long form. Which means, that color presets might break.

**Comment by BWPanda** [comment]
@indigoxela Could you please add your feedback, and maybe an explanation, here then? https://github.com/backdrop-ops/api.backdropcms.org/issues/50

If we can get that issue actioned first, I can change this one to suit.

**Comment by BWPanda** [comment]
I've made a PR to fix Color module and make it properly support short hex codes: https://github.com/backdrop/backdrop-issues/issues/4913

If that gets support and is merged, then theme authors can adhere to our coding standards and we don't need to change them 😃

**Comment by BWPanda** [comment]
I have updated the [Stylelint rules PR](https://github.com/backdrop/backdrop/pull/3516) to change CSS colors from short to long as per the recent decision to change the CSS coding standards in that regard. However I can't recall how I ran the stylelint rules against core in order to update the second PR, so that'll have to wait until I have time to look into this again...

**Comment by BWPanda** [comment]
I've updated the PR that adds Stylelint rules to core (removing the deprecated stylistic ones): https://github.com/backdrop/backdrop/pull/3516
And the PR that applies these CSS standards to core and fixes a bunch of stuff: https://github.com/backdrop/backdrop/pull/4026
